errors.go

```diff
@@ -1,52 +1,96 @@
 	errMakeRequestError          = "error from makeRequest"
 	errUnmarshalError            = "error unmarshalling the JSON response"
+	errUnmarshalErrorBody        = "error unmarshalling the JSON response error body"
 	errRequestNotSuccessful      = "error reported by API"
 	errMissingAccountID          = "account ID is empty and must be provided"
 	errOperationStillRunning     = "bulk operation did not finish before timeout"
 	errOperationUnexpectedStatus = "bulk operation returned an unexpected status"
 )
 
-var _ Error = &UserError{}
+// APIRequestError is a type of error raised by API calls made by this library.
+type APIRequestError struct {
+	StatusCode int
+	Errors     []ResponseInfo
+}
+
+func (e APIRequestError) Error() string {
+	errString := ""
+	errString += fmt.Sprintf("HTTP status %d: ", e.StatusCode)
+
+	errMessages := []string{}
+	for _, err := range e.Errors {
+		m := ""
+		if err.Message != "" {
+			m += fmt.Sprintf("%s", err.Message)
+		}
+
+		if err.Code != 0 {
+			m += fmt.Sprintf(" (%d)", err.Code)
+		}
+
+		errMessages = append(errMessages, m)
+	}
 
-// Error represents an error returned from this library.
-type Error interface {
-	error
-	// Raised when user credentials or configuration is invalid.
-	User() bool
-	// Raised when a parsing error (e.g. JSON) occurs.
-	Parse() bool
-	// Raised when a network error occurs.
-	Network() bool
-	// Contains the most recent error.
+	return errString + strings.Join(errMessages, ", ")
 }
 
-// UserError represents a user-generated error.
-type UserError struct {
-	Err error
+// HTTPStatusCode exposes the HTTP status from the error response encountered.
+func (e APIRequestError) HTTPStatusCode() int {
+	return e.StatusCode
 }
 
-// User is a user-caused error.
-func (e *UserError) User() bool {
-	return true
+// ErrorMessages exposes the error messages as a slice of strings from the error
+// response encountered.
+func (e *APIRequestError) ErrorMessages() []string {
+	messages := []string{}
+
+	for _, e := range e.Errors {
+		messages = append(messages, e.Message)
+	}
+
+	return messages
+}
+
+// InternalErrorCodes exposes the internal error codes as a slice of int from
+// the error response encountered.
+func (e *APIRequestError) InternalErrorCodes() []int {
+	ec := []int{}
+
+	for _, e := range e.Errors {
+		ec = append(ec, e.Code)
+	}
+
+	return ec
 }
 
-// Network error.
-func (e *UserError) Network() bool {
-	return false
+// ServiceError returns a boolean whether or not the raised error was caused by
+// an internal service.
+func (e *APIRequestError) ServiceError() bool {
+	return e.StatusCode >= http.StatusInternalServerError
 }
 
-// Parse error.
-func (e *UserError) Parse() bool {
-	return true
+// ClientError returns a boolean whether or not the raised error was caused by
+// something client side.
+func (e *APIRequestError) ClientError() bool {
+	return e.StatusCode >= http.StatusBadRequest &&
+		e.StatusCode <= http.StatusInternalServerError &&
```

Comment:
shouldn't this be `<` instead of `<=`?

Reply:
good catch, it is! I'll push a fix next week for this.

Reply:
updated in 0afe83f